File: AsciiNumberOcr.hs
module AsciiNumberOcr (translate) where

import Data.List(transpose)
import Data.List.Split(chunksOf)

translate :: String -> String
translate " _ | ||_|" = "0"
translate "     |  |" = "1"
translate " _  _||_ " = "2"
translate " _  _| _|" = "3"
translate "   |_|  |" = "4"
translate " _ |_  _|" = "5"
translate " _ |_ |_|" = "6"
translate " _   |  |" = "7"
translate " _ |_||_|" = "8"
translate " _ |_| _|" = "9"
translate s = concatMap (translate . concat) $ transpose $ map (chunksOf 3) $ lines s

File: asciiNumberOcr.js
const numbersMap = [
  ' _ | ||_|   ',
  '     |  |   ',
  ' _  _||_    ',
  ' _  _| _|   ',
  '   |_|  |   ',
  ' _ |_  _|   ',
  ' _ |_ |_|   ',
  ' _   |  |   ',
  ' _ |_||_|   ',
  ' _ |_| _|   '
]

module.exports = (asciiNumber) => {
  const lines = asciiNumber.split('\n')

  const chars = lines.map(line => {
    return line.match(/.{3}/g)
  }).filter(line => line != null)

  const transposed = transpose(chars)

  const charArray = transposed.map(chars => numbersMap.indexOf(chars.join('')))

  return charArray.join('')
}

function transpose(rows) {
  const temp = [];

  for (let r = 0; r < rows.length; ++r) {
    for (let c = 0; c < rows[r].length; ++c) {
      if (!temp[c]) {
        temp[c] = [];
      }

      temp[c][r] = rows[r][c];
    }
  }

  return temp;
}